torchvision/prototype/datasets/_builtin/semeion.py

```diff
@@ -0,0 +1,69 @@
+import io
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import torch
+from torchdata.datapipes.iter import (
+    IterDataPipe,
+    Mapper,
+    Shuffler,
+    CSVParser,
+)
+from torchvision.prototype.datasets.decoder import raw
+from torchvision.prototype.datasets.utils import (
+    Dataset,
+    DatasetConfig,
+    DatasetInfo,
+    HttpResource,
+    OnlineResource,
+    DatasetType,
+)
+from torchvision.prototype.datasets.utils._internal import INFINITE_BUFFER_SIZE
+
+
+class SEMEION(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "semeion",
+            type=DatasetType.RAW,
+            categories=10,
+            homepage="https://archive.ics.uci.edu/ml/datasets/Semeion+Handwritten+Digit",
+        )
+
+    def resources(self, config: DatasetConfig) -> List[OnlineResource]:
+        archive = HttpResource(
+            "http://archive.ics.uci.edu/ml/machine-learning-databases/semeion/semeion.data",
+            sha256="f43228ae3da5ea6a3c95069d53450b86166770e3b719dcc333182128fe08d4b1",
+        )
+        return [archive]
+
+    def _collate_and_decode_sample(
+        self,
+        data: Tuple[str, ...],
+        *,
+        decoder: Optional[Callable[[io.IOBase], torch.Tensor]],
+    ) -> Dict[str, Any]:
+        image_data = torch.tensor([float(pixel) for pixel in data[:256]]).reshape(16, 16)
+        label_data = [int(label) for label in data[256:] if label]
+
+        if decoder is raw:
+            image = image_data.unsqueeze(0)
+        else:
+            image_buffer = image_buffer_from_array(image_data.numpy())
+            image = decoder(image_buffer) if decoder else image_buffer  # type: ignore[assignment]
```

Comment:
Currently [`mypy` complains](https://app.circleci.com/pipelines/github/pytorch/vision/12173/workflows/449bfc48-cc58-4b67-a2d7-0b98d978f7e8/jobs/956162) that the `type: ignore` directive is unused. I'd wager a guess that this warning goes away when we fix the import above. Try running `mypy` before you remove it.

Reply:
Thanks, @pmeier for taking a look. I ran `mypy --config=mypy.ini torchvision/prototype/datasets/_builtin/semeion.py` and it doesn't throw an error now (after your fix suggestion).

But, interestingly, the script to test it's functioning doesn't work on my system:

```python
from torchvision.prototype import datasets

for sample in datasets.load('semeion', decoder=datasets.decoder.pil):
    print(sample)
    break
```

It throws an `OSError: cannot write mode F as png`:

```python
Traceback (most recent call last):
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/PIL/PngImagePlugin.py", line 1229, in _save
    rawmode, mode = _OUTMODES[mode]
KeyError: 'F'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/krshrimali/copy/vision/main.py", line 3, in <module>
    for sample in datasets.load('semeion', decoder=datasets.decoder.pil):
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/torch/utils/data/datapipes/iter/callable.py", line 144, in __iter__
    yield self._apply(data, self.nesting_level)
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/torch/utils/data/datapipes/iter/callable.py", line 120, in _apply
    return self._apply_fn(data)
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/torch/utils/data/datapipes/iter/callable.py", line 85, in _apply_fn
    return self.fn(data, *self.args, **self.kwargs)
  File "/home/krshrimali/copy/vision/torchvision/prototype/datasets/_builtin/semeion.py", line 51, in _collate_and_decode_sample
    image_buffer = image_buffer_from_array(image_data.numpy())
  File "/home/krshrimali/copy/vision/torchvision/prototype/datasets/utils/_internal.py", line 181, in image_buffer_from_array
    image.save(buffer, format=format)
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/PIL/Image.py", line 2240, in save
    save_handler(self, fp, filename)
  File "/home/krshrimali/.conda/envs/original-torch/lib/python3.9/site-packages/PIL/PngImagePlugin.py", line 1231, in _save
    raise OSError(f"cannot write mode {mode} as PNG") from e
OSError: cannot write mode F as PNG
```

I'll take a look at why this might happen, but if it's something that you have seen before or know a solution to - I'll appreciate any pointers. :) Thanks!

Reply:
Update: looks like the issue is resolved if the array is converted to `dtype: np.uint8`. Will send a fix soon.

Reply:
See https://github.com/pytorch/vision/pull/4840#discussion_r741654240.

Reply:
Should be fixed now. No `mypy` issues on my system, so marking this as resolved. Please let me know if there is anything else that I missed. Thanks for the help! :)